### DISCUSSION TITLE
ElevatedManagedAction scheduled before RemoveExistingProducts alternative

### DISCUSSION THREAD

**Discussion by Trigve** [oleg-shilo/wixsharp#1704] [Q&A] [answered]
Hi,
I have a problem with my msi when doing upgrade. When major upgrade is run I have a custom `ManagedAction` run before `RemoveExistingProducts` to stop some services before the upgrade (uninstall of the old version). The problem is that in some circumstances, some of those actions could fail, when trying to stop the service with `ServiceController` with error message of `Acces is denied`. So it looks like permission problem (What is weird is that some services will stop without error and some not). So I've tried the quick fix to change the action to `ElevatedManagedAction` but then then installer will error out, because elevated actions couldn't be run before `RemoveExistingProducts.

So what are my options then?

**Comment by oleg-shilo** [comment, marked as answer]
You can always run an elevated process from the unelevated CA:

```C#
[CustomAction]
public static ActionResult InvokeRemoveFiles(Session session)
{
    try
    {
        var startInfo = new ProcessStartInfo();

        startInfo.UseShellExecute = true;
        startInfo.FileName = "sc";
        startInfo.Arguments = "stop <ServiceName>";
        startInfo.Verb = "runas";

        Process.Start(startInfo);

        return ActionResult.Success;
    }
    catch(Exception e)
    {
        // handle the error here
    }
}
```

**Reply by Trigve** [reply]
Thanks for the reply.
I was hoping for some "pure" wixsharp solution but after trying a lot of things it looks like using the process would be the best one.

**Reply by oleg-shilo** [reply]
That's the thing. WiX is a wonderful product but it is limited to the concept it is based on - declarative syntax for relationships between MSI entities and a very opinionated API that reflects the madness of MSI database schema.
This is exactly why WixSharp was created. To extend WiX with the possibility to create setup "behaviour"
 without no limits by leveraging the full power of a proper programming language.
